Content.Server/MassDriver/EntitySystems/MassDriverSystem.cs

```diff
@@ -0,0 +1,336 @@
+            return;
+
+        var hasComp = HasComp<ActiveMassDriverComponent>(uid);
+        if (hasComp && !args.Powered)
+            RemComp<ActiveMassDriverComponent>(uid);
+        else if (!hasComp && args.Powered)
+            EnsureComp<ActiveMassDriverComponent>(uid);
+    }
+
+    /// <summary>
+    /// Update only active mass drivers, so we have it more optimized than conveyors
+    /// </summary>
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<ActiveMassDriverComponent, MassDriverComponent, ApcPowerReceiverComponent>();
+        while (query.MoveNext(out var uid, out var activeMassDriver, out var massDriver, out var powered))
+        {
+            if (_timing.CurTime < activeMassDriver.NextUpdateTime)
+                continue;
+
+            if (activeMassDriver.NextUpdateTime == TimeSpan.Zero)
+            {
+                activeMassDriver.NextUpdateTime = _timing.CurTime + activeMassDriver.UpdateDelay;
+                continue;
+            }
+            activeMassDriver.NextUpdateTime = _timing.CurTime + activeMassDriver.UpdateDelay;
+
+            if (!_powerReceiver.IsPowered(uid))
+                continue;
+
+            var entities = new HashSet<EntityUid>();
+            _lookup.GetEntitiesIntersecting(uid, entities);
+            var entitiesCount = entities.Count(a => !Transform(a).Anchored);
+
+            if (entitiesCount == 0)
+            {
+                // Disable mass driver if we throw all entities
+                if (activeMassDriver.NextThrowTime != TimeSpan.Zero)
+                {
+                    if (TryComp<AmbientSoundComponent>(uid, out var ambient))
+                        _audioSystem.SetAmbience(uid, false, ambient);
+                    activeMassDriver.NextThrowTime = TimeSpan.Zero;
+                    _appearance.SetData(uid, MassDriverVisuals.Launching, false);
+                    _powerReceiver.SetLoad(powered, massDriver.MassDriverPowerLoad);
+                }
+                if (massDriver.Mode == MassDriverMode.Manual)
+                    RemComp<ActiveMassDriverComponent>(uid);
+                continue;
+            }
+
+            // If we find first entity, charge mass driver(wait n seconds setuped in ThrowDelay)
+            if (activeMassDriver.NextThrowTime == TimeSpan.Zero)
+            {
+                activeMassDriver.NextThrowTime = _timing.CurTime + massDriver.ThrowDelay;
+                continue;
+            }
+            else if (_timing.CurTime < activeMassDriver.NextThrowTime)
+                continue;
+
+            _powerReceiver.SetLoad(powered, massDriver.LaunchPowerLoad);
+            _appearance.SetData(uid, MassDriverVisuals.Launching, true);
+
+            ThrowEntities(uid, massDriver, entities, entitiesCount);
+
+            if (TryComp<AmbientSoundComponent>(uid, out var ambientSound))
+                _audioSystem.SetAmbience(uid, true, ambientSound);
```

Comment:
This is really going to hurt performance for a few reasons:
1. It's checking poweredcomp every time when it should just be active / inactive based on power state changes (see above).
2. It has LINQ inside.
3. It allocates the entity hashset for every single massdriver.
4. None of the timers are networked when they should be so the throwing can be predicted on the client so it looks good.